fusequery/query/src/optimizers/optimizer_projection_push_down.rs

```diff
@@ -0,0 +1,289 @@
+use common_planners::ReadDataSourcePlan;
+
+use crate::optimizers::IOptimizer;
+use crate::optimizers::Optimizer;
+use crate::sessions::FuseQueryContextRef;
+
+pub struct ProjectionPushDownOptimizer {}
+
+impl ProjectionPushDownOptimizer {
+    pub fn create(_ctx: FuseQueryContextRef) -> Self {
+        ProjectionPushDownOptimizer {}
+    }
+}
+
+/// Recursively walk an expression tree, collecting the unique set of column names
+/// referenced in the expression
+fn expr_to_column_names(expr: &ExpressionPlan, accum: &mut HashSet<String>) -> Result<()> {
+    let expressions = Optimizer::expression_plan_children(expr)?;
+
+    let _expressions = expressions
+        .iter()
+        .map(|e| expr_to_column_names(e, accum))
+        .collect::<Result<Vec<_>>>()?;
+
+    if let ExpressionPlan::Column(name) = expr {
+        accum.insert(name.clone());
+    }
+    Ok(())
+}
+
+/// Recursively walk a list of expression trees, collecting the unique set of column
+/// names referenced in the expression
+fn exprvec_to_column_names(expr: &[ExpressionPlan], accum: &mut HashSet<String>) -> Result<()> {
+    for e in expr {
+        expr_to_column_names(e, accum)?;
+    }
+    Ok(())
+}
+
+fn expr_to_name(e: &ExpressionPlan) -> Result<String> {
+    match e {
+        ExpressionPlan::Column(name) => Ok(name.clone()),
+        _ => Err(anyhow::anyhow!("Ignore ExpressionPlan that is not Column.")),
+    }
+}
+
+fn get_projected_schema(
+    schema: &DataSchema,
+    required_columns: &HashSet<String>,
+    has_projection: bool,
+) -> Result<DataSchemaRef> {
+    // Discard non-existing columns, e.g. when the column derives from aggregation
+    let mut projection: Vec<usize> = required_columns
+        .iter()
+        .map(|name| schema.index_of(name))
+        .filter_map(ArrowResult::ok)
+        .collect();
+    if projection.is_empty() {
+        if has_projection {
+            // Ensure reading at lease one column
+            projection.push(0);
+        } else {
+            // for table scan without projection
+            // just return all columns
+            projection = schema
+                .fields()
+                .iter()
+                .enumerate()
+                .map(|(i, _)| i)
+                .collect::<Vec<usize>>();
+        }
+    }
+    // sort the projection to get deterministic behavior
+    projection.sort_unstable();
+
+    // create the projected schema
+    let mut projected_fields: Vec<DataField> = Vec::with_capacity(projection.len());
+    for i in &projection {
+        projected_fields.push(schema.fields()[*i].clone());
+    }
+    Ok(Arc::new(DataSchema::new(projected_fields)))
+}
+
+fn optimize_plan(
+    optimizer: &ProjectionPushDownOptimizer,
+    plan: &PlanNode,
+    required_columns: &HashSet<String>,
+    has_projection: bool,
+) -> Result<PlanNode> {
+    let mut new_required_columns = required_columns.clone();
+    println!("(1) {:?}", required_columns);
+    match plan {
```

Comment:
`SELECT a FROM table_name WHERE b = LITERAL ` how does it work?

Reply:
Do you have any idea about how I should write a unit test for this statement?

Reply:
https://github.com/datafuselabs/datafuse/pull/286/files#r615336014  is helpful.